src/sentry/lang/javascript/processor.py

```diff
@@ -695,26 +714,189 @@ def format_groups(p):
+        newest entry.
+        """
+        # TODO: in the future we would like to load all the artifact bundles that are connected to the projects
+        #  we have permissions on not all the bundles.
+        return DebugIdArtifactBundle.objects.filter(
+            organization_id=self.organization.id,
+            debug_id=debug_id,
+            source_file_type=source_file_type.value if source_file_type is not None else None,
+        ).select_related("artifact_bundle__file")[0]
+
+    @staticmethod
+    def _fetch_artifact_bundle_file(artifact_bundle):
+        """
+        Fetches the File object bound to an ArtifactBundle.
+
+        The File object represents the actual .zip bundle file which contains all the required artifacts.
+        """
+        # We try to load the bundle file from the cache. The file we are loading here is the entire bundle, not
+        # the contents.
+        cache_key = get_artifact_bundle_cache_key(artifact_bundle.bundle_id)
+        result = cache.get(cache_key)
+        if result:
+            return BytesIO(result)
+
+        # We didn't find the bundle in the cache, thus we want to fetch it.
+        artifact_bundle_file = fetch_retry_policy(artifact_bundle.file.getfile)
+
+        # `cache.set` will only keep values up to a certain size,
+        # so we should not read the entire file if it's too large for caching
+        if CACHE_MAX_VALUE_SIZE is not None and artifact_bundle_file.size > CACHE_MAX_VALUE_SIZE:
+            return artifact_bundle_file
+
+        with sentry_sdk.start_span(op="_fetch_artifact_bundle_file.read_for_caching") as span:
+            span.set_data("file_size", artifact_bundle_file.size)
+            contents = artifact_bundle_file.read()
+        with sentry_sdk.start_span(op="_fetch_artifact_bundle_file.write_to_cache") as span:
+            span.set_data("file_size", len(contents))
+            cache.set(cache_key, contents, 3600)
+
+        artifact_bundle_file.seek(0)
+        return artifact_bundle_file
+
+    def _open_artifact_bundle_archive(self, debug_id, source_file_type):
+        """
+        Opens an ArtifactBundle as a .zip file and returns an ArtifactBundleArchive object that allows the caller
+        to read index individual files within the archive.
+
+        This function uses two levels of caching:
+        1. self.open_archives is a local cache that persists for the Fetcher lifetime and contains all the opened
+        archives indexed by bundle_id.
+        2. self._fetch_artifact_bundle_file uses inside the default cache which is hosted on memcached and stores the
+        actual File object bound to a specific ArtifactBundle. memcached is persisted across processor runs as opposed
+        to the local cache.
+        """
+        bundle_id = None
+
+        try:
+            # We first have to run a query to determine the bundle_id that contains the tuple
```

Comment:
just as a NOTE:
As I implemented this in symbolicator, I basically made this a try-fetch-retry operation:

First look into all the `open_archives` for the file. If none is found, query the API for the archives that contain the file, then iterate over `open_archives` again.

That would potentially avoid API requests for debug-ids that are already present in fetched archives.

Reply:
Yeah, that could also be a worthwhile implementation, considering that we expect to have not many `open_archives` and they are already loaded into memory. It will save us from several expensive multi-join queries that resolve to most likely the same `bundle_id`.